internal/verifier/change_stream.go

```diff
@@ -103,31 +113,55 @@ func (verifier *Verifier) initializeChangeStreamReaders() {
 		changeStreamRunning:  false,
 		changeEventBatchChan: make(chan []ParsedEvent),
 		writesOffTs:          util.NewEventual[primitive.Timestamp](),
-		error:                util.NewEventual[error](),
+		readerError:          util.NewEventual[error](),
+		handlerError:         util.NewEventual[error](),
 		doneChan:             make(chan struct{}),
 		lag:                  msync.NewTypedAtomic(option.None[time.Duration]()),
 	}
 }
 
-// StartChangeEventHandler starts a goroutine that handles change event batches from the reader.
-// It needs to be started after the reader starts.
-func (verifier *Verifier) StartChangeEventHandler(ctx context.Context, reader *ChangeStreamReader) error {
-	for {
+// RunChangeEventHandler handles change event batches from the reader.
+// It needs to be started after the reader starts and should run in its own
+// goroutine.
+func (verifier *Verifier) RunChangeEventHandler(ctx context.Context, reader *ChangeStreamReader) error {
+	var err error
+
+	for err == nil {
 		select {
 		case <-ctx.Done():
-			return ctx.Err()
+			err = util.WrapCtxErrWithCause(ctx)
+
+			verifier.logger.Debug().
+				Err(err).
+				Stringer("changeStreamReader", reader).
+				Msg("Closing change event handler.")
+
 		case batch, more := <-reader.changeEventBatchChan:
 			if !more {
```

Comment:
This isn't related to this PR, but it occurs to me that this may be a bug. I'm pretty sure that `batch` can be non-empty even when `more` is false.

Reply:
I think that’s incorrect: if `more` is false, `batch` is always the zero value.

[Go by Example](https://gobyexample.com/closing-channels)

Reply:
Ah, I misunderstood some writing about channels by Dave Cheney, I think.